concent_api/core/models_helpers.py

```diff
@@ -0,0 +1,38 @@
+from typing import Any
+from typing import Optional
+from typing import Type
+from typing import Union
+
+from django.db import IntegrityError
+from django.db.models import Model, QuerySet
+from django.db.models.base import ModelBase
+from psycopg2 import errorcodes as pg_errorcodes
+
+
+def get_one_or_none(
+    subtask_or_query_set: Union[Model, QuerySet],
+    **conditions: Any
+)-> Optional[Model]:
+    if isinstance(subtask_or_query_set, ModelBase):
+        instances = subtask_or_query_set.objects.filter(**conditions)
+        assert len(instances) <= 1
+        return None if len(instances) == 0 else instances[0]
+    else:
+        instances = subtask_or_query_set.filter(**conditions)
+        assert len(instances) <= 1
+        return None if len(instances) == 0 else instances[0]
+
+
+def get_or_create_safely(model: Type[Model], **kwargs: Any) -> Model:
+    try:
+        model_object = model.objects.get_or_create_full_clean(
+            **kwargs,
+        )
+    except IntegrityError as exception:
```

Comment:
since we use `AUTOCOMMIT=True` it will be sufficient not to have this code in any `atomic` block.

Reply:
In Your version of `settle_overdue_acceptances` there is lot of short `atomic` blocks. In general transactions are to make some changes and commit all or nothing. In this case, when we want to commit one by one (change->commit, change->commit) I'm not sure that transactions are what we want. If You would call `get_or_create_safely` outside of transaction, it will call database queries inside of its own transactions anyway (django wraps every database query in its own transaction). I think it will work exactly as @cameel wrote, commit will be before `except` and the solution is simple to remove ` with transaction.atomic(using='control'):` outside of `get_or_create_safely`

Reply:
> since we use `AUTOCOMMIT=True` it will be sufficient not to have this code in any `atomic` block.

What do you mean by this? We do not use autocommit. We have Django set up to use `ATOMIC_REQUESTS` which means that every view is wrapped in a transaction by default, even if you don't use `@atomic` explicitly. And in many other places (Celery tasks, management commands) we put the whole function in a transaction.

If anyone calls this function from within an `@atomic` block (and that's how we **always** use `get_one_or_none()`), the change will not be committed inside the `try ... except` block.

Reply:
> In this case, when we want to commit one by one (change->commit, change->commit) I'm not sure that transactions are what we want.

We always want database queries to be wrapped in transactions. Actually there's no way not to. Even autocommit is just a single-query transaction that's committed immediately.